examples_tests/02.ComputeShader/main.cpp

```diff
@@ -27,11 +38,584 @@ void main()
+			descriptorPoolSizeCount, poolSizes);
+
+	// For each swapchain image we have one descriptor set with two descriptors each
+	core::smart_refctd_ptr<video::IGPUDescriptorSet> descriptorSets[MAX_SWAPCHAIN_IMAGE_COUNT];
+
+	// Todo(achal): Test this as well: 
+	// device->createGPUDescriptorSets(descriptorPool.get(), SC_IMG_COUNT, )
+	for (uint32_t i = 0u; i < swapchainImageCount; ++i)
+	{
+		descriptorSets[i] = device->createGPUDescriptorSet(descriptorPool.get(),
+			core::smart_refctd_ptr(dsLayout));
+	}
+
+	core::smart_refctd_ptr<video::IGPUBuffer> ubos[MAX_SWAPCHAIN_IMAGE_COUNT];
+
+	// Kinda janky to pass buffer size like this, I think, because we're overriding
+	// these values at the end of createGPUBuffer anyway (by the values obtained from
+	// vkGetBufferMemoryRequirements)
+	video::IDriverMemoryBacked::SDriverMemoryRequirements memoryRequirements = {};
+	memoryRequirements.vulkanReqs.size = sizeof(UniformBufferObject);
+	memoryRequirements.vulkanBufferUsageFlags = VK_BUFFER_USAGE_UNIFORM_BUFFER_BIT;
+	memoryRequirements.sharingMode = asset::ESM_EXCLUSIVE;
+	memoryRequirements.queueFamilyIndices = nullptr;
+	for (uint32_t i = 0u; i < swapchainImageCount; ++i)
+		ubos[i] = device->createGPUBuffer(memoryRequirements);
+
+	// Allocate memory for GPU buffer
+	core::smart_refctd_ptr<video::IDriverMemoryAllocation> ubosMemory[MAX_SWAPCHAIN_IMAGE_COUNT];
+	for (uint32_t i = 0u; i < swapchainImageCount; ++i)
+	{
+		// THIS IS ONLY USED FOR DETERMINING THE SIZE OF ALLOCATION FOR NOW
+		// Again kinda janky to pass size of allocation this way
+		video::IDriverMemoryBacked::SDriverMemoryRequirements additionalMemReqs = {};
+		additionalMemReqs.vulkanReqs.size = sizeof(UniformBufferObject);
+		additionalMemReqs.vulkanReqs.memoryTypeBits = ubos[i]->getMemoryReqs().vulkanReqs.memoryTypeBits;
+
+		ubosMemory[i] = device->allocateDeviceLocalMemory(additionalMemReqs);
+	}
+
+	video::ILogicalDevice::SBindBufferMemoryInfo bindBufferInfos[MAX_SWAPCHAIN_IMAGE_COUNT];
+	for (uint32_t i = 0u; i < swapchainImageCount; ++i)
+	{
+		bindBufferInfos[i].buffer = ubos[i].get();
+		bindBufferInfos[i].memory = ubosMemory[i].get();
+		bindBufferInfos[i].offset = 0ull;
+	}
+	device->bindBufferMemory(swapchainImageCount, bindBufferInfos);
+
+	// Fill up ubos with dummy data
+	// Todo(achal): Would probably make sense to ditch map/unmap in favor of staging buffer,
+	// infact device local memory shouldn't be host mapped
+	struct UniformBufferObject uboData_cpu[3] = { { 1.f, 0.f, 0.f, 1.f}, {0.f, 1.f, 0.f, 1.f}, {0.f, 0.f, 1.f, 1.f} };
+	for (uint32_t i = 0u; i < swapchainImageCount; ++i)
+	{
+		video::IDriverMemoryAllocation::MappedMemoryRange mappedMemoryRange(ubosMemory[i].get(),
+			0ull, sizeof(UniformBufferObject));
+		void* mappedMemoryAddress = device->mapMemory(mappedMemoryRange);
+		assert(mappedMemoryAddress);
+
+		memcpy(mappedMemoryAddress, &uboData_cpu[i], sizeof(UniformBufferObject));
+
+		device->unmapMemory(ubosMemory[i].get());
+	}
```

Comment:
make flushing, etc work and the staging buffer will work

Reply:
resolved